### PR TITLE
LDAP: Divide the requests

### PR DESCRIPTION
**Pull request by markelog**
Active Directory does indeed have a limitation with 1000 results
per search (default of course).

However, that limitation can be workaround with the pagination search feature,
meaning `pagination` number is how many times LDAP compatible server will be
requested by the client with specified amount of users (like 1000). That feature
already embeded with LDAP compatible client (including our `go-ldap`).

But slapd server has by default stricter settings. First, limitation is not 1000
but 500, second, pagination workaround presumably (information about it a bit
scarce and I still not sure on some of the details from my own testing)
cannot be workaround with pagination feature.

See
https://www.openldap.org/doc/admin24/limits.html
https://serverfault.com/questions/328671/paging-using-ldapsearch
hashicorp/vault#4162 - not sure why they were hitting the limit in
the first place, since `go-ldap` doesn't have one by default.

But, given all that, for me `ldapsearch` command with same request
as with `go-ldap` still returns more then 500 results, it can even return
as much as 10500 items (probably more).

So either there is some differences with implementation of the LDAP search
between `go-ldap` module and `ldapsearch` or I am missing a step :/.

In the wild (see serverfault link), apparently, people still hitting that
limitation even with `ldapsearch`, so it still seems to be an issue.

But, nevertheless, I'm still confused by this incoherence.

To workaround it, I divide the request by no more then 500 items per search

### CODE REVIEW

File: pkg/services/ldap/ldap.go [hunk 1]
```diff
@@ -148,10 +154,78 @@ func (server *Server) Login(query *models.LoginUserQuery) (
...
+	return nil
+}
+
 // Users gets LDAP users
 func (server *Server) Users(logins []string) (
 	[]*models.ExternalUserInfo,
 	error,
+) {
+	var users []*ldap.Entry
+	err := getUsersIteration(logins, func(previous, current int) error {
```
Comment (markelog) [inline pkg/services/ldap/ldap.go]: As a side note, these requests could be done in parallel, hm, maybe in another iteration?

**Review by alexanderzobnin — APPROVED** [review]
Looks good to me!

File: pkg/services/ldap/ldap.go [hunk 2]
```diff
@@ -148,10 +154,78 @@ func (server *Server) Login(query *models.LoginUserQuery) (
...
+// It divides the users by equal parts for the anticipated requests
+func getUsersIteration(logins []string, fn func(int, int) error) error {
+	lenLogins := len(logins)
+	iterations := int(
+		math.Ceil(
+			float64(lenLogins) / float64(UsersMaxRequest),
+		),
+	)
+
+	for i := 1; i < iterations+1; i++ {
```
Comment (xlson) [inline pkg/services/ldap/ldap.go]: I think we could simplify this loop to something like this:

```go
	for i := 1; i < iterations+1; i++ {
		previous := math.Max(0, float64(UsersMaxRequest*(i-1)))
		current := math.Min(float64(i*UsersMaxRequest), float64(lenLogins))

		err := fn(int(previous), int(current))
		if err != nil {
			return err
		}
	}
```

File: pkg/services/ldap/ldap.go (hunk 1 above)
Comment (xlson) [reply pkg/services/ldap/ldap.go]: Good idea. I think we can start off by running it like this and parallelize it later if needed.

**Review by xlson — APPROVED** [review]
Great job!

File: pkg/services/ldap/ldap.go (hunk 2 above)
Comment (markelog) [reply pkg/services/ldap/ldap.go]: ooh, I love it!

Comment (markelog) [reply pkg/services/ldap/ldap.go]: Managed to simplified it even further :)

Comment (xlson) [reply pkg/services/ldap/ldap.go]: Neat!

**Pull request merged**